### ISSUE TITLE
Optimize the size of the executable

### ISSUE THREAD

**Issue by user1823** [theJayTea/WritingTools#9]
First of all, thanks for this great software. It is indeed very useful.

However, I feel that the size of the executable is too large for a software that just sends a prompt to an online service (Gemini) and fetches the result.

Is it possible to reduce the size of the executable somehow? For example, by reducing the number of dependencies or by substituting some of the dependencies by slimmer alternatives (like replacing PySide6 with PySide6-Essentials)?

**Comment by Nisekoi-1** [comment]
Yeah, 100 MB of RAM is a bit much.

**Comment by theJayTea** [comment]
Hi @user1823 !

You're right, it does use more RAM than I'd like (although the CPU usage is optimised).

I slimmed down the PySide6 implementation with a new Build File, and tried a whole lot of optimisations with it but I can't seem to get it down below 100 MB myself without using something like UPX (which results in antivirus false positives).

I'd welcome any contributions you can make to optimising the file size!

**Comment by user1823** [comment]
Sorry, but I don't have the necessary expertise to help you with this.

But, maybe you can try using `--onedir` instead of `--onefile` with PyInstaller locally and see which component occupies the largest amount of disk space, and then try to optimize that.

Anyway, congratulations for reducing the file size by ~10 MB while including additional functionality.

**Comment by theJayTea** [comment]
I tried both of them lol, they both made it larger (~150-200 MB) XD

It's the libraries that occupy the most space, and I've stripped them as much as I currently could.

And haha thanks.

**Comment by Nisekoi-1** [comment]
Would it be possible to implement a feature where the app enters 'sleep' mode to use less RAM while running in the background? When assigned keys are pressed, the app would 'wake up,' albeit with a slightly slower response speed. It would remain active for 5-10 minutes before going back to 'sleep'. This could be made an optional toggle for users who prefer to keep the app running continuously with faster response times.

**Comment by user1823** [comment]
> Would it be possible to implement a feature where the app enters 'sleep' mode to use less RAM while running in the background? 

I think Windows does that automatically. In the screenshot below, see that Writing Tools is using just 3.6 MB of RAM.

![image](https://github.com/user-attachments/assets/e9d0c579-4240-4c8a-a0e2-6efcff5552c1)

The current issue is aimed at reducing the size of the .exe file.

**Comment by Vadim-Karpenko** [comment]
Building from source results in 66.8MB for me. The only difference is that I use virtualenv to make sure I have only the required stuff installed and nothing else.
![image](https://github.com/user-attachments/assets/d24de42c-8271-4aed-9837-67ebd6222e86)

Try the following
```
pip install virtualenv
virtualenv myvenv
```

Then, install the requirements and build the file. It should make a difference.

**Comment by uninstall-your-browser** [comment]
Maybe you could use nuitka. It could also reduce the start up time for the application too. The [documentation](https://nuitka.net/user-documentation/user-manual.html) is a little bit hard to follow, but it has a plugin for pyside (`--enable-plugin=pyside6`)

You might want to use the `pyside6-essential` package which excludes alot of the stuff that you are excluding in the pyinstaller build script.

Also, I noticed you talked about packaging for Linux on reddit. The only cross-distro way I found to do that with a pyside app is to make it a flatpak.

**Comment by theJayTea** [comment]
@Vadim-Karpenko , wow. I did use a venv but I don’t think i did it optimally/correctly, so I’ll work on this for the next update and re-compilation. Thanks for sharing that!

**Comment by theJayTea** [comment]
@uninstall-your-browser , I did initially use Nuitka (the first release uses it!) but it led to constant antivirus false positives, unfortunately. This is a difficult choice.

I looked into PySide6-Essentials, but swapping `from PySide6 import...` to `from PySide6-Essentialls` import did not work at all.

I don't have a perfect grasp on how they've structured this, but it seems that the main `pyside-6` is already just the essentials, with the additional stuff as `pyside-6 addons` (which we're not using)?

Regardless, in the build script, I explicitly excluded almost all the other PySide6 stuff that we aren't using.

**Comment by uninstall-your-browser** [comment]
> I looked into PySide6-Essentials, but swapping `from PySide6 import...` to `from PySide6-Essentials` import did not work at all.

I don't think you need to change anything, it just excludes some extra stuff like the web browser view

Were the AV problems with building it or running it? Because I think it does something funny with injecting resources into the exe when building, which is something that viruses typically do.

**Comment by theJayTea** [comment]
Nuitka caused problems when running it. This is a known issue with it.

The reason is that it obfuscated and minifies the code a ton (to make it smaller), which is also something that malware does to hide its functionality.

In addition, it converts the Python code to C and uses a C compiler, further obfuscating what's going on.

It was so bad that even adding exclusions in my AV software didn't reliably work, across PCs and AVs haha.

**Comment by theJayTea** [comment]
I tried using PySide6-Essentials again. While it resulted in a 45mb exe, the program didn't work with it (the pop-up never showed up) :(

**Comment by nightkall** [comment]
> Building from source results in 66.8MB for me. The only difference is that I use virtualenv to make sure I have only the required stuff installed and nothing else. ![image](https://private-user-images.githubusercontent.com/26556210/378224841-d24de42c-8271-4aed-9837-67ebd6222e86.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3MDAzMzcsIm5iZiI6MTcyOTcwMDAzNywicGF0aCI6Ii8yNjU1NjIxMC8zNzgyMjQ4NDEtZDI0ZGU0MmMtODI3MS00YWVkLTk4MzctNjdlYmQ2MjIyZTg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIzVDE2MTM1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMwMDI3NTlhODgzYjVhODc5Y2U4ZDU5NGE1MjYyZjNiMTZhOWM2MjdlZTY4MjU5ZTQwMDAwMDdiNmY1MmNjM2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.H-78L5JFTuwBMOr7DeaW0L3ABg344DKP-_rOVevrzTg)
> 
> Try the following
> 
> ```
> pip install virtualenv
> virtualenv myvenv
> myvens\Scripts\activate
> ```
> 
> Then, install the requirements and build the file. It should make a difference.

I tried that
```
pip install virtualenv
virtualenv myvenv
myvenv\Scripts\activate
```
While in (myvenv), then:

```
cd path\to\Windows_and_Linux
pip install -r requirements.txt
pip install pyinstaller
pythonw pyinstaller-build-script.py
```

The compiled .exe is in \WritingTools\Windows_and_Linux\dist\Writing Tools.exe
and is **62,8 MB (65.936.818 bytes)**

![rfV5JGaT3H](https://github.com/user-attachments/assets/793126f1-c1f3-4952-92ac-975e55bc48c8)

So yes, it's smaller but it has more VirusTotal antivirus positives than the official release:

Official [Writing.Tools.v4.zip](https://github.com/theJayTea/WritingTools/releases/download/v4/Writing.Tools.v4.zip) [VirusTotal](https://www.virustotal.com/gui/file/21d7693f15d5c5a3b1797f800bed76cae5bbd1f1803d91984d318de382eda864/detection) report:
![NpPmT78d3T](https://github.com/user-attachments/assets/35cc8c05-e594-49cd-b3ed-197ae6e2b7b6)


My compiled [version](https://github.com/theJayTea/WritingTools/commit/ed5248e3aea4a12c790e9f555d4167a95bcf5da5) [VirusTotal](https://www.virustotal.com/gui/file/9f5d589b9125296c94e64b997de0eabea13355a2019f026dbca8eb9a42b100ef/detection) report:
![VCYW9KjKUi](https://github.com/user-attachments/assets/65f45408-e46f-4c60-9bce-1c666c94cecc)

**Comment by theJayTea** [comment]
Hey @nightkall, it's great to know that it compiled to a much smaller file size for you. I don't think I used the venv correctly, and PyInstaller must've included a lot more installed packages for me (I'd first installed pyside6-essentials and it didn't work, and then installed the whole pyside 6 _again_ in my venv so that may have added to bloat?). Regardless, for the next release I'll compile it in my VM inside a venv to just get this optimised haha.

About the memory use, it doesn't actually use up ~500MB, that's comitted memory (not in use, just the OS saying that it _could_ use that much if it wanted to). 

<img width="434" alt="image" src="https://github.com/user-attachments/assets/36098685-f81f-4da3-9a62-9c6a3e847ba9">


It's quite normal for committed memory to report higher than the memory actually in use - Task Manager:
![image](https://github.com/user-attachments/assets/08346a6a-0818-40e1-ab92-14bda8750108)

With that said, I'm sure we can optimise memory use further down the line, as it is a bit much.

**Comment by nightkall** [comment]
> About the memory use, it doesn't actually use up ~500MB, that's comitted memory (not in use, just the OS saying that it _could_ use that much if it wanted to).
> It's quite normal for committed memory to report higher than the memory actually in use - Task Manager:

You're right about the committed memory, I forgot to show Working set column. 


> It's quite normal for committed memory to report higher than the memory actually in use - Task Manager:
> About the threads, that's curious. Did you launch both of them around the same time? I'm trying to rule out it being a minor memory leak.

I just launched both at the same time and your compiled version has more threads than mine by default, without invoking any commands.

Here is an updated screenshot: 

![Writing Tools Mem-Threads](https://github.com/user-attachments/assets/60320959-75bc-43a6-ba6e-3671ba522264)

**Comment by theJayTea** [comment]
I’m happy to share that I’ve finally been able to reduce the size of the executable by more than half with the latest release :D 
It took a lot of experimentation, and it turned out that my venvs weren’t working correctly before.
